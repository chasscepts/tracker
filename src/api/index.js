import axios from 'axios';

const normalizeError = (err) => {
  if (!err) {
    return { message: 'An unknown error encountered. Please try again.' };
  }
  if (err.response) {
    return { message: err.response.data.message || JSON.stringify(err.response.data) };
  }
  if (err.request) {
    return { message: 'Server is not responding.' };
  }
  if (err.message) {
    return { message: err.message };
  }
  if (typeof err === 'string') {
    return { message: err };
  }
  return { message: 'An unknown error encountered. Please try again.' };
};

const instantiate = (token) => {
  const instance = axios.create({
    baseURL: 'http://localhost:3000/',
    responseType: 'json',
  });
  instance.defaults.headers.common.Authorization = `Bearer ${token}`;

  return instance;
};

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const get = (instance, path) => new Promise((resolve, reject) => {
  instance.get(path)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @param {{ string: any }} data sent in body of this post
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const post = (instance, path, data) => new Promise((resolve, reject) => {
  instance.post(path, data)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

const api = {
  getGroups: (token) => get(instantiate(token), '/groups'),
  getTasks: (token, start, end) => {
    let path = '/tasks';
    if (start) {
      path = `${path}?start=${start}`;
      if (end) {
        path = `${path}&end=${end}`;
      }
    }

    return get(instantiate(token), path);
  },
  updateTask: (token, title) => post(instantiate(token), '/tasks', { title }),
};

export default api;
