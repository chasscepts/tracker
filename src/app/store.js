import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authSlice';
import tasksReducer from '../reducers/tasksSlice';
import timerReducer from '../reducers/timerSlice';

export default configureStore({
  reducer: {
    auth: authReducer,
    tasks: tasksReducer,
    timer: timerReducer,
  },
});
