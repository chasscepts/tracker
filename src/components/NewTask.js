import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { COLORS } from '../utilities';
import button from '../assets/css/button.module.css';
import {
  selectCreateTaskError,
  selectCreateTaskRequestCount,
  setCreateTaskError,
  selectGroups,
  loadGroups,
  createTask,
} from '../reducers/tasksSlice';
import LdsRing from './LdsRing';

const styles = {
  container: {
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    overflow: 'hidden',
  },
  inner: {
    width: '100%',
    maxWidth: '350px',
    border: '1px solid #ddd',
    borderRadius: '5px',
    padding: '15px',
    overflow: 'auto',
  },
  h2: {
    marginBottom: '15px',
    color: COLORS.primary,
    borderBottom: `1px dotted ${COLORS.primary}`,
    paddingBottom: '5px',
  },
  label: {
    fontWeight: 'bold',
    fontSize: '0.9rem',
    marginBottom: '3px',
  },
  input: {
    display: 'block',
    width: '100%',
    padding: '7px',
    border: '1px solid #ddd',
    marginBottom: '15px',
    color: 'inherit',
  },
  controls: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
  loadingWrap: {
    padding: '5px',
    display: 'flex',
    alignItems: 'center',
    border: '1px solid #eee',
    boxShadow: '0 0 1px 1px',
    borderRadius: '5px',
    marginBottom: '15px',
  },
  loadingMsg: {
    marginLeft: '10px',
  },
  error: {
    position: 'relative',
    fontWeight: 'bold',
    fontSize: '0.8rem',
    padding: '8px',
    color: COLORS.red,
    border: '1px solid #ddd',
    textAlign: 'center',
    marginBottom: '15px',
    maxHeight: '150px',
    overflow: 'auto',
  },
  errorBtn: {
    position: 'absolute',
    right: '3px',
    top: '3px',
    padding: 0,
    border: 'none',
    color: COLORS.red,
    backgroundColor: 'transparent',
    cursor: 'pointer',
  },
};

export default function NewTask() {
  const [groupId, setGroupId] = useState('');
  const [title, setTitle] = useState('');
  const errorMsg = useSelector(selectCreateTaskError);
  const requestCount = useSelector(selectCreateTaskRequestCount);
  const groups = useSelector(selectGroups);
  const dispatch = useDispatch();

  let options = [];
  if (groups) {
    options = groups;
    if (!groupId) {
      setGroupId(options[0].id);
    }
  } else {
    setTimeout(() => dispatch(loadGroups()), 0);
  }

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    if (name === 'id') {
      setGroupId(value);
    } else if (name === 'title') {
      setTitle(value);
    }
  };

  const clearError = () => dispatch(setCreateTaskError(null));

  const submitForm = () => {
    if (groupId && title) {
      dispatch(createTask(groupId, title));
    }
  };

  const loadingMessage = requestCount > 1 ? `${requestCount} requests ongoing` : `${requestCount} request ongoing`;

  return (
    <div style={styles.container}>
      <div style={styles.inner}>
        <h2 style={styles.h2}>New Task</h2>
        {requestCount > 0
        && (
        <div style={styles.loadingWrap}>
          <LdsRing width={20} color={COLORS.primary} />
          <div style={styles.loadingMsg}>{loadingMessage}</div>
        </div>
        )}
        {errorMsg
        && (
        <div style={styles.error}>
          <button style={styles.errorBtn} type="button" onClick={clearError}>X</button>
          <div>{errorMsg}</div>
        </div>
        )}
        <div style={styles.label}>Please Select Task Category</div>
        <select style={styles.input} name="id" value={groupId} onChange={handleChange}>
          {options.map((g) => <option key={g.title} value={g.id}>{g.title}</option>)}
        </select>
        <div style={styles.label}>Enter Title</div>
        <input style={styles.input} name="title" value={title} onChange={handleChange} />
        <div style={styles.controls}>
          <button
            className={`${button.btn} ${button.blue} ${button.round}`}
            type="button"
            onClick={submitForm}
          >
            Create Task
          </button>
        </div>
      </div>
    </div>
  );
}
