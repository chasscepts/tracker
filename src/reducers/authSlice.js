import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

/* eslint-disable no-param-reassign */
const authSlice = createSlice({
  name: 'auth',
  initialState: {
    authenticated: false,
    loginError: null,
    registrationError: null,
    token: null,
    registrationSuccess: false,
  },
  reducers: {
    setToken: (state, { payload }) => {
      state.authenticated = true;
      state.token = payload;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
    setLoginError: (state, { payload }) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = payload;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
    setRegistrationError: (state, { payload }) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = payload;
      state.registrationSuccess = false;
    },
    setRegistrationSuccess: (state) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = true;
    },
    logout: (state) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  setToken,
  setLoginError,
  setRegistrationError,
  setRegistrationSuccess,
  logout,
} = authSlice.actions;

export const login = (email, password) => (dispatch) => {
  const url = 'http://localhost/auth';
  axios({
    url,
    method: 'post',
    responseType: 'json',
    data: { email, password },
  })
    .then((res) => {
      dispatch(setToken(res.data.token));
    })
    .catch((err) => {
      if (!err) {
        dispatch(setLoginError('Unknown error encountered. Please try again.'));
        return;
      }
      if (err.response) {
        dispatch(setLoginError(err.response.data));
        return;
      }
      if (err.request) {
        dispatch(setLoginError('Server is not responding.'));
        return;
      }
      if (err.message) {
        dispatch(setLoginError(err.message));
        return;
      }
      if (typeof err === 'string') {
        dispatch(setLoginError(err));
        return;
      }
      dispatch(setLoginError('Unknown error encountered. Please try again.'));
    })
    .catch((err) => {
      console.log(err);
    });
};

export const register = (email, password) => (dispatch) => {
  const url = 'http://localhost/auth/register';
  axios({
    url,
    method: 'post',
    responseType: 'json',
    data: { email, password },
  })
    .then(() => {
      dispatch(setRegistrationSuccess());
    })
    .catch((err) => {
      if (!err) {
        dispatch(setRegistrationError('An unknown error encountered. Please try again.'));
        return;
      }
      if (err.response) {
        dispatch(setRegistrationError(err.response.data));
        return;
      }
      if (err.request) {
        dispatch(setRegistrationError('Server is not responding.'));
        return;
      }
      if (err.message) {
        dispatch(setRegistrationError(err.message));
        return;
      }
      if (typeof err === 'string') {
        dispatch(setRegistrationError(err));
        return;
      }
      dispatch(setRegistrationError('An unknown error encountered. Please try again.'));
    })
    .catch((err) => {
      console.log(err);
    });
};

export const selectAuthenticated = (state) => state.auth.authenticated;

export const selectLoginError = (state) => state.auth.loginError;

export const selectRegistrationError = (state) => state.auth.registrationError;

export default authSlice.reducer;
