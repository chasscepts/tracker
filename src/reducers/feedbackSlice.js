import { createSlice } from '@reduxjs/toolkit';

/* eslint-disable no-param-reassign */
const feedbackSlice = createSlice({
  name: 'feedback',
  initialState: {
    id: 0,
    feedbacks: [],
  },
  reducers: {
    addFeedback: (state, { payload }) => {
      state.feedbacks.push({ id: state.id, payload });
      state.id += 1;
    },
    removeFeedback: (state, { payload }) => {
      state.feedbacks = state.feedbacks.filter((f) => f.id !== payload);
    },
  },
});
/* eslint-enable no-param-reassign */

export const { addFeedback, removeFeedback } = feedbackSlice.actions;

export const selectFeedbacks = (state) => state.feedback.feedbacks;

export default feedbackSlice.reducer;
