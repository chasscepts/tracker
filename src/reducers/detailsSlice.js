import { createSlice } from '@reduxjs/toolkit';
import api from '../api';

/* eslint-disable no-param-reassign */
const detailsSlice = createSlice({
  name: 'details',
  initialState: {
    requestCount: 0,
    hasPendingRequest: false,
    store: [],
    group: null,
    loadError: null,
  },
  reducers: {
    addRequest: (state) => {
      state.requestCount += 1;
      state.hasPendingRequest = true;
    },
    addGroup: (state, { payload }) => {
      state.group = payload;
      state.store.push(payload);
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
    setGroup: (state, { payload }) => {
      state.group = payload;
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
    setGroupLoadError: (state, { payload }) => {
      state.loadError = payload;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  addRequest,
  addGroup,
  setGroup,
  setGroupLoadError,
} = detailsSlice.actions;

export const selectDetailsHasPendingError = (state) => state.details.hasPendingRequest;
export const selectDetailsGroup = (state) => state.details.group;
export const selectDetailsLoadError = (state) => state.details.loadError;

export const loadGroupTasks = (groupId, title) => (dispatch, getState) => {
  const { auth: { user }, details: { store } } = getState();
  if (!user) {
    dispatch(setGroupLoadError('You must be logged in to fetch tasks'));
    return;
  }

  const group = store.find((g) => g.id === groupId);
  if (group) {
    dispatch(setGroup(group));
    return;
  }

  dispatch(addRequest());

  api.getGroupTasks(user.token, groupId)
    .then((group) => {
      dispatch(addGroup({ title, id: groupId, tasks: group }));
    })
    .catch((err) => setGroupLoadError(err.message));
};

export default detailsSlice.reducer;
