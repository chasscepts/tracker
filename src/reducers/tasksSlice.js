import { createSlice } from '@reduxjs/toolkit';
import {
  fetchGroups, fetchTasks, createTask, updateTask, deleteTask,
} from '../api';
import { today, daysAgo } from '../utilities/dates';
import { addFeedback } from './feedbackSlice';

/* eslint-disable no-param-reassign */
const tasksSlice = createSlice({
  name: 'tasks',
  initialState: {
    date: today(),
    groups: null,
    groupsLoading: false,
    tasksCache: {},
    tasks: [],
    tasksLoading: false,
    createRequestCount: 0,
    updateRequestCount: 0,
  },
  reducers: {
    setDate: (state, { payload }) => {
      state.date = payload;
    },
    setGroups: (state, { payload }) => {
      state.groups = payload;
      state.groupsLoading = false;
    },
    setGroupsLoading: (state, { payload }) => {
      state.groupsLoading = payload;
    },
    setTasks: (state, { payload }) => {
      state.tasks = payload;
    },
    setTasksLoading: (state, { payload }) => {
      state.tasksLoading = payload;
    },
    updateTasksCache: (state, { payload: { date, tasks } }) => {
      state.tasksCache[date] = tasks;
    },
    addCreateTaskRequest: (state) => {
      state.createRequestCount += 1;
    },
    addTask: (state, { payload }) => {
      state.tasks.push(payload);
      if (state.createRequestCount > 0) {
        state.createRequestCount -= 1;
      }
    },
    removeCreateRequest: (state) => {
      if (state.createRequestCount > 0) {
        state.createRequestCount -= 1;
      }
    },
    addUpdateTaskRequest: (state) => {
      state.updateRequestCount += 1;
    },
    removeUpdateTaskRequest: (state) => {
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    updateLocalTask: (state, { payload: { id, title } }) => {
      const task = state.tasks.find((t) => t.id === id);
      if (task) {
        task.title = title;
      }
      Object.keys(state.tasksCache).forEach((key) => {
        const task = state.tasksCache[key].find((t) => t.id === id);
        if (task) {
          task.title = title;
        }
      });
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    deleteLocalTask: (state, { payload }) => {
      state.tasks = state.tasks.filter((t) => t.id !== payload);
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    updateTaskEntry: (state, { payload: { entry, duration } }) => {
      const task = state.tasks.find((t) => t.id === entry.task_id);
      if (task) {
        task.entries[0].duration += duration;
      }
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  setDate,
  setGroups,
  setGroupsLoading,
  setTasks,
  setTasksLoading,
  updateTasksCache,
  addTask,
  addCreateTaskRequest,
  removeCreateRequest,
  updateLocalTask,
  addUpdateTaskRequest,
  removeUpdateTaskRequest,
  deleteLocalTask,
  updateTaskEntry,
} = tasksSlice.actions;

export const selectDate = (state) => state.tasks.date;

export const selectGroups = (state) => state.tasks.groups;

export const selectGroupsLoading = (state) => state.tasks.groupsLoading;

export const selectTasks = (state) => state.tasks.tasks;

export const selectCreateTaskRequestCount = (state) => state.tasks.createRequestCount;

export const selectUpdateTaskRequestCount = (state) => state.tasks.updateRequestCount;

export const loadGroups = () => (dispatch, getState) => {
  const state = getState();
  if (state.tasks.groups || state.tasks.groupsLoading) return;

  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch groups', type: 'error' }));
    return;
  }

  dispatch(setGroupsLoading(true));

  fetchGroups(user.token)
    .then((groups) => {
      dispatch(setGroups(groups));
    })
    .catch(({ message }) => {
      dispatch(setGroupsLoading(false));
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const loadTasks = () => (dispatch, getState) => {
  const state = getState();

  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch tasks!', type: 'error' }));
    return;
  }

  dispatch(setTasksLoading());

  fetchTasks(user.token, { start: daysAgo(14) })
    .then((tasks) => {
      dispatch(setTasks(tasks));
    })
    .catch(({ message }) => dispatch(addFeedback({ message, type: 'error' })));
};

export const createTaskAsync = (groupId, title) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to create tasks', type: 'error' }));
    return;
  }

  dispatch(addCreateTaskRequest());

  createTask(user.token, groupId, title)
    .then((task) => {
      dispatch(addTask(task));
      dispatch(addFeedback({ message: `The task ${task.title} was successfully created.`, type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeCreateRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const updateTaskAsync = (id, title) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to update tasks', type: 'error' }));
    return;
  }

  dispatch(addUpdateTaskRequest());

  updateTask(user.token, id, title)
    .then(() => {
      dispatch(updateLocalTask({ id, title }));
      dispatch(addFeedback({ message: `Task ${title} successfully updated.`, type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeUpdateTaskRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const deleteTaskAsync = (id) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to update tasks', type: 'error' }));
    return;
  }

  dispatch(addUpdateTaskRequest());

  deleteTask(user.token, id)
    .then(() => {
      dispatch(deleteLocalTask(id));
      dispatch(addFeedback({ message: 'Task successfully deleted.', type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeUpdateTaskRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export default tasksSlice.reducer;
