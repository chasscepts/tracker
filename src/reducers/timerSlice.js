import { createSlice } from '@reduxjs/toolkit';
import api from '../api';
import { setTasksError } from './tasksSlice';

/* eslint-disable no-param-reassign */
const timerSlice = createSlice({
  name: 'timer',
  initialState: {
    entry: {},
    nextEntry: {},
    updateError: null,
    updateSuccess: null,
    pendingRequestCount: 0,
    hasPendingRequests: false,
  },
  reducers: {
    commitNext: (state) => {
      state.entry = state.nextEntry;
      state.nextEntry = null;
    },
    setEntry: (state, { payload }) => {
      state.entry = payload;
    },
    setNextEntry: (state, { payload }) => {
      if (!state.entry.id) {
        state.entry = payload;
      } else {
        state.nextEntry = payload;
      }
    },
    addEntryRequest: (state) => {
      state.pendingRequestCount += 1;
      state.hasPendingRequests = true;
      state.updateError = null;
      state.updateSuccess = null;
    },
    updateStoreEntry: (state, { payload: { message, duration } }) => {
      console.log({ message, duration });
      state.updateSuccess = message;
      state.entry.entry.duration += duration;
      console.log(state.entry.entry.duration);
      let temp = state.pendingRequestCount - 1;
      if (temp <= 0) {
        state.hasPendingRequests = false;
        temp = 0;
      }
      state.pendingRequestCount = temp;
    },
    setEntryUpdateSuccess: (state, { payload }) => {
      state.updateSuccess = payload;
      let temp = state.pendingRequestCount - 1;
      if (temp <= 0) {
        state.hasPendingRequests = false;
        temp = 0;
      }
      state.pendingRequestCount = temp;
    },
    setEntryUpdateError: (state, { payload }) => {
      state.updateError = payload;
      let temp = state.pendingRequestCount - 1;
      if (temp <= 0) {
        state.hasPendingRequests = false;
        temp = 0;
      }
      state.pendingRequestCount = temp;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  commitNext,
  setEntry,
  setNextEntry,
  addEntryRequest,
  updateStoreEntry,
  setEntryUpdateSuccess,
  setEntryUpdateError,
} = timerSlice.actions;

export const selectEntry = (state) => state.timer.entry;
export const selectNextEntry = (state) => state.timer.nextEntry;
export const selectEntryUpdateError = (state) => state.timer.updateError;
export const selectEntryUpdateSuccess = (state) => state.timer.updateSuccess;
export const selectEntryHasPendingRequest = (state) => state.timer.hasPendingRequests;
export const selectEntryPendingRequestCount = (state) => state.timer.pendingRequestCount;

export default timerSlice.reducer;

export const updateEntry = (entry, duration, title) => (dispatch, getState) => {
  const { auth: { user } } = getState();

  if (!user) {
    dispatch(setTasksError('You must be logged in to fetch tasks'));
    return;
  }

  dispatch(addEntryRequest());

  api.updateEntry(user.token, entry, duration)
    .then(() => {
      const msg = `Update of Task Entry ${title} successful`;
      dispatch(updateStoreEntry({ message: msg, duration }));
    })
    .catch(({ message }) => {
      const msg = `Update of Task Entry ${title} failed with message: ${message}`;
      dispatch(setEntryUpdateError(msg));
    });
};
