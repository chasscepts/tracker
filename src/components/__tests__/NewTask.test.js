import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Wrapper from '../../test_helpers/wrapper';
import NewTask from '../NewTask';
import { createTask } from '../../api';
import { createStore } from '../../app/store';
import login from '../../test_helpers/login';
import { setGroups } from '../../reducers/tasksSlice';

jest.mock('../../api');

const groups = [{ id: 1, title: 'Group 1' }];

describe('NewTask', () => {
  it('matches snapshot', async () => {
    const store = createStore();
    store.dispatch(setGroups(groups));
    const tree = renderer.create(<Wrapper store={store} Component={NewTask} />);
    expect(tree.toJSON()).toMatchSnapshot();
    tree.unmount();
  });

  it('calls api to create new task with correct parameter', async () => {
    const store = createStore();
    login(store);
    store.dispatch(setGroups(groups));
    const title = 'Title';

    createTask.mockResolvedValue({ id: 1, title });

    render(<Wrapper store={store} Component={NewTask} />);
    userEvent.type(screen.getByPlaceholderText('Enter Title'), title);
    await userEvent.click(screen.getByText('Create Task'));
    await Promise.resolve();
    expect(createTask).toHaveBeenCalledWith(expect.any(String), expect.any(Number), title);
  });
});
