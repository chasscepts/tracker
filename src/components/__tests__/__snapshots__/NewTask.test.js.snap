// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTask matches snapshot 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "maxWidth": "400px",
      "overflow": "hidden",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "border": "1px solid #ddd",
        "borderRadius": "5px",
        "maxWidth": "350px",
        "overflow": "auto",
        "padding": "15px",
        "width": "100%",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <h2
        style={
          Object {
            "borderBottom": "1px dotted #62b5e5",
            "color": "#62b5e5",
            "marginBottom": "15px",
            "paddingBottom": "5px",
          }
        }
      >
        New Task
      </h2>
      <div
        style={
          Object {
            "fontSize": "0.9rem",
            "fontWeight": "bold",
            "marginBottom": "3px",
          }
        }
      >
        Please Select Task Category
      </div>
      <select
        name="id"
        onChange={[Function]}
        style={
          Object {
            "border": "1px solid #ddd",
            "color": "inherit",
            "display": "block",
            "marginBottom": "15px",
            "padding": "7px",
            "width": "100%",
          }
        }
        value={1}
      >
        <option
          value={1}
        >
          Group 1
        </option>
      </select>
      <div
        style={
          Object {
            "fontSize": "0.9rem",
            "fontWeight": "bold",
            "marginBottom": "3px",
          }
        }
      >
        Enter Title
      </div>
      <input
        name="title"
        onChange={[Function]}
        placeholder="Enter Title"
        style={
          Object {
            "border": "1px solid #ddd",
            "color": "inherit",
            "display": "block",
            "marginBottom": "15px",
            "padding": "7px",
            "width": "100%",
          }
        }
        value=""
      />
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "flex-end",
          }
        }
      >
        <button
          className="btn blue round"
          type="submit"
        >
          Create Task
        </button>
      </div>
    </form>
  </div>
</div>
`;
